### PR TITLE
delete expunge

### PR DESCRIPTION
**Pull request by fil512**

### CODE REVIEW

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java [hunk 1]
```diff
@@ -916,6 +916,95 @@
...
+	 * java.lang.String - Contains the url used to delete and expunge the resources
+	 * </li>
+	 * </ul>
+	 * <p>
+	 * Hooks should return <code>void</code>. They may choose to throw an exception however, in
+	 * which case the delete expunge will not occur.
+	 * </p>
+	 */
+
+	STORAGE_DELETE_EXPUNGE_PID_LIST(
```
Comment (jamesagnew) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java:999]: I'm not sure about the name of this.. The pointcuts are almost all named after steps in the processing lifecycle. We should give this a name  that makes it clear how this is different from `STORAGE_PRE_DELETE_EXPUNGE` (if it is? can they be collapsed?)

File: hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/rest/client/GenericClientDstu3Test.java [hunk 2]
```diff
@@ -1187,7 +1196,9 @@ public InputStream answer(InvocationOnMock theInvocation) throws Throwable {
 		idx++;
 	}
 
+	// FIXME KHS disable until it's fixed
 	@Test
+	@Disabled
```
Comment (jamesagnew) [inline hapi-fhir-structures-dstu3/src/test/java/ca/uhn/fhir/rest/client/GenericClientDstu3Test.java:1201]: Fix has been merged to master

```suggestion
```

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpDelete.java [hunk 3]
```diff
@@ -30,4 +30,9 @@
 	 */
 	IAuthRuleBuilderRuleOp onCascade();
 
+	/**
+	 * Specifies that this rule applies to delete expunges as opposed to regular
+	 * deletes.  A delete expunge is a delete operation called with the _expunge=true parameter.
+	 */
+	IAuthRuleBuilderRuleOp onExpunge();
```
Comment (jamesagnew) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpDelete.java:37]: Is there a test for this? I'd recommend putting one or two in AuthorizationInterceptorJpaR4Test

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java:999]: One is pre all of them, the other receives the pid chunks

Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java:999]: Changed to STORAGE_PRE_DELETE_EXPUNGE_PID_LIST

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpDelete.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpDelete.java:37]: thanks.  there's a test in a different place, but I like the tests here; I have added the tests.

**Pull request merged**